File: js/src/PassBookProgram.ts
import { PublicKey } from '@solana/web3.js';
import { Program } from '@metaplex-foundation/mpl-core';
import { Store } from './accounts';

export class PassBookProgram extends Program {
  static readonly PREFIX = 'passbook';
  static readonly PUBKEY = new PublicKey('passK9sjcBkUzWu35gf2x4EmpcrkZB9NXgHWtgAzxhB');

  static async findProgramAuthority(): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [Buffer.from(PassBookProgram.PREFIX, 'utf8'), PassBookProgram.PUBKEY.toBuffer()],
      PassBookProgram.PUBKEY,
    );
  }

  static async findPassStoreAccount(authority: PublicKey): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from(PassBookProgram.PREFIX, 'utf8'),
        PassBookProgram.PUBKEY.toBuffer(),
        authority.toBuffer(),
        Buffer.from(Store.STORE_PREFIX, 'utf8'),
      ],
      PassBookProgram.PUBKEY,
    );
  }

  static async findPassBookAccount(mint: PublicKey): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from(PassBookProgram.PREFIX, 'utf8'),
        PassBookProgram.PUBKEY.toBuffer(),
        mint.toBuffer(),
      ],
      PassBookProgram.PUBKEY,
    );
  }

  static async findPassAccount(mint: PublicKey): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from(PassBookProgram.PREFIX, 'utf8'),
        PassBookProgram.PUBKEY.toBuffer(),
        mint.toBuffer(),
      ],
      PassBookProgram.PUBKEY,
    );
  }
}

File: dart/lib/passbook_program.dart
import 'package:passbook/accounts/store.dart';
import 'package:solana/encoder.dart';
import 'package:solana/solana.dart';

class PassbookProgram {
  static const prefix = 'passbook';
  static const programId = 'passjvPvHQWN4SvBCmHk1gdrtBvoHRERtQK9MKemreQ';

  static Future<String> findProgramAuthority() {
    return findProgramAddress(seeds: [
      Buffer.fromBase58(prefix),
      Buffer.fromBase58(PassbookProgram.programId),
    ], programId: programId);
  }

  static Future<String> findPassStoreAccount(String authority) {
    return findProgramAddress(seeds: [
      Buffer.fromBase58(prefix),
      Buffer.fromBase58(PassbookProgram.programId),
      Buffer.fromBase58(authority),
      Buffer.fromBase58(Store.prefix),
    ], programId: programId);
  }

  static Future<String> findPassBookAccount(String mint) {
    return findProgramAddress(seeds: [
      Buffer.fromBase58(prefix),
      Buffer.fromBase58(PassbookProgram.programId),
      Buffer.fromBase58(mint),
    ], programId: programId);
  }

  static Future<String> findPassAccount(String mint) {
    return findProgramAddress(seeds: [
      Buffer.fromBase58(prefix),
      Buffer.fromBase58(PassbookProgram.programId),
      Buffer.fromBase58(mint),
    ], programId: programId);
  }
}
